### PR TITLE
[megatron] feat: Support MTP training in SFT 

### PR DESCRIPTION
**Pull request by arvyanh**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

The SFT training with MTP is supported, using the same MTP training configuration as RL training.

An example configuration for running SFT can be found in `examples/sft/gsm8k/run_mimo_megatron_mtp.sh`

**SFT result**

The experiment was conducted using following data:
- model = mimo-7B-math
- dataset = gsm8k

The result: [wandb link](https://wandb.ai/hou-zg-meituan/mimo-7b-sft-mtp?nw=nwuserhouzg)

The presence of mtp layer has limited effect on main loss. However, when MTP layer is detached, the mtp_loss converges to a higher value.

### Checklist Before Starting

- [X] Search for similar PRs. Paste at least one query link here: ...
- [X] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### Design & Code Changes

Support SFT training with MTP, with the core change based on https://github.com/volcengine/verl/pull/4936

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [X] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [X] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [X] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [X] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: Using none-standard mbridge/mcore
- [X] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [X] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

File: verl/models/mcore/model_forward.py [hunk 1]
```diff
@@ -233,6 +243,19 @@ def gptmodel_forward_no_padding(
...
+        if enable_mtp and post_process:
+            args = {
+                k: preprocess_bshd_no_padding(v, pre_process=True, need_roll=(k == "label" or k == "loss_mask"))[0]
+                for k, v in logits_processor_args.items()
+            }
+            model_kwargs["labels"] = args["label"].contiguous()
+            model_kwargs["loss_mask"] = args["loss_mask"].contiguous()
+        logits_processor_args.pop("loss_mask")
+
+        # print(f'GPTModel fwd: pre_process={pre_process}, post_process={post_process}')
```
Comment (ArronHZG) [inline verl/models/mcore/model_forward.py]: remove print log

File: verl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -370,9 +382,9 @@ def fit(self):
                         metric = {"val/loss": val_loss.detach().item()}
                         tracking.log(data=metric, step=global_step)
                         last_valid_metric = metric
-                    torch.distributed.barrier()
```
Comment (ArronHZG) [inline verl/trainer/sft_trainer.py:373]: why this barrier be removed?

File: verl/utils/dataset/multiturn_sft_dataset.py [hunk 3]
```diff
@@ -416,4 +448,8 @@ def sanity_check(self, input_ids: torch.Tensor, messages: list[dict], tools: lis
             if self.ignore_input_ids_mismatch:
                 logger.warning_once(error_message)
             else:
+                print(input_ids)
+                print()
+                print()
```
Comment (ArronHZG) [inline verl/utils/dataset/multiturn_sft_dataset.py]: remove print log

File: verl/utils/dataset/multiturn_sft_dataset.py [hunk 4]
```diff
@@ -125,7 +149,8 @@ def series_to_item(ls):
 
         dataframes = []
         for parquet_file in self.parquet_files:
-            dataframe = pd.read_parquet(parquet_file)
+            # default loader loads some list as np.ndarray, which fails the tokenizer
+            dataframe = pd.read_parquet(parquet_file, dtype_backend="pyarrow")
```
Comment (ArronHZG) [inline verl/utils/dataset/multiturn_sft_dataset.py:158]: why pb need pyarrrow params?

File: verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+from verl.utils.dataset.multiturn_sft_dataset import MultiTurnSFTDataset
+
+
+class MultiTurnSFTDatasetDeepseek(MultiTurnSFTDataset):
+    def tokenize_assistant(self, index, message, full_message, tools, enable_thinking):
+        """
+        reimplement the jinja logic to suit multiturn data for dsv31
+        """
+
+        # 判断当前assistant消息是否有tool_calls
```
Comment (ArronHZG) [inline verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py]: use english Annotation

File: verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                        + formatted_args
+                        + "<｜tool▁call▁end｜>",
+                        add_special_tokens=False,
+                    )
+            tokens += processor.encode("    <｜tool▁calls▁end｜><｜end▁of▁sentence｜>", add_special_tokens=False)
+        else:
+            content = message.get("content", "")
+            tokens += processor.encode(prefix + content + "<｜end▁of▁sentence｜>", add_special_tokens=False)
+
+        # 构造inputs
```
Comment (ArronHZG) [inline verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py]: use english Annotation

File: verl/workers/engine/megatron/transformer_impl.py [hunk 7]
```diff
@@ -710,13 +720,16 @@ def forward_step(self, batch_iter, model, postprocess_micro_batch_func):
 
         forward_fn = get_mcore_forward_no_padding_fn(self.model_config.hf_config)
 
+        has_mtp = self.model_config.mtp.enable
+
```
Comment (ArronHZG) [inline verl/workers/engine/megatron/transformer_impl.py]: redundant code

File: verl/workers/engine_workers.py [hunk 8]
```diff
@@ -354,6 +360,42 @@ def save_checkpoint(self, local_path, hdfs_path=None, global_step=0, max_ckpt_to
     def load_checkpoint(self, local_path, hdfs_path=None, del_local_after_load=False):
         return self.engine.load_checkpoint(local_path, hdfs_path, del_local_after_load)
 
+    def check_engine_config(self):
```
Comment (ArronHZG) [inline verl/workers/engine_workers.py]: change func name for mtp

File: verl/trainer/sft_trainer.py [hunk 9]
```diff
@@ -149,7 +155,9 @@ def _init_engine(self):
         if self.test_freq == "after_each_epoch":
             self.test_freq = self.steps_per_epoch
 
+        self.training_client.check_engine_config()
```
Comment (ArronHZG) [inline verl/trainer/sft_trainer.py]: move this func to megatron impl

File: verl/utils/megatron_utils.py [hunk 10]
```diff
@@ -1242,6 +1243,28 @@ def mapping_string_to_attn_backend(args: dict) -> dict:
     return args
 
 
+def get_megatron_mtp_loss(output, n_micro_batch):
```
Comment (ArronHZG) [inline verl/utils/megatron_utils.py]: reuse get_megatron_mtp_loss  in verl rl process

File: verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
+# Copyright 2025 Meituan Ltd. and/or its affiliates
```
Comment (vermouth1992) [inline verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py]: I would suggest that instead of writing a new class for DeepSeek, we can make `_process_single_message` a function that can be loaded as a function inside a module to keep the codebase clean

Comment (arvyanh) [reply verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py]: I'm a bit unsure about what the difference is between loading a function and loading a class (that only contains that function) in terms of making the codebase clean?

Are you suggesting that we should only have only single top-level class for all templates?

Comment (vermouth1992) [reply verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py]: I assume that there will be a lot of customization of process_single_item function for different models. All the other logics are shared,  especially if we implement streaming dataset to handle large files.

Comment (arvyanh) [reply verl/utils/dataset/multiturn_sft_dataset_deepseek_v31.py]: I think I'll move this to a separate PR. I wanted to align 1:1 with sglang's deepseekV3 chat_template (and could pass MultiTurnSFTDataset's sanity checks), the current status is that:
- Still failing in some cases
- Too many template-specific (manually adding spaces and tokens) tweaks, which doesn't looks nice
I'll probably add some abstraction first

File: verl/workers/engine/megatron/transformer_impl.py [hunk 12]
```diff
@@ -262,7 +263,44 @@ def is_mp_src_rank_with_outputs(self):
             and mpu.get_context_parallel_rank() == 0
         )
 
+    def check_mtp_config(self):
```
Comment (ISEEKYAN) [inline verl/workers/engine/megatron/transformer_impl.py]: I advise to move this function to another place to keep the model engine implementation clean

File: verl/trainer/config/sft_trainer_engine.yaml [hunk 13]
```diff
@@ -26,6 +26,7 @@ data:
   messages_key: messages  # Key for messages list in multi-turn mode
   tools_key: tools  # Key for tools list in multi-turn mode
   enable_thinking_key: enable_thinking  # Whether to enable thinking in multi-turn mode
+  enable_thinking_default: none         # The default value when enable_thinking_key is not present in the dataset
```
Comment (ISEEKYAN) [inline verl/trainer/config/sft_trainer_engine.yaml:29]: why is this label related to MTP feature?

**Review by ISEEKYAN — APPROVED** [review]

**Review by wuxibin89 — APPROVED** [review]

**Review by ArronHZG — APPROVED** [review]

**Comment by vermouth1992** [comment]
Shall we create a simple CI for this using SFT? Otherwise, it may be very easy to be break.. We can use a tiny random weight init model to perform CI

**Comment by Eisenhower** [comment]
@arvyanh @ISEEKYAN Is there a plan to support / adapt MiMo-V2-Flash?
